// wangct-util/lib/typeUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isFunc = isFunc;
exports.isAsyncFunc = isAsyncFunc;
exports.isNum = isNum;
exports.isArray = isArray;
exports.isAry = isAry;
exports.isBoolean = isBoolean;
exports.isBol = isBol;
exports.isObject = isObject;
exports.isObj = isObj;
exports.isString = isString;
exports.isStr = isStr;
exports.isDate = isDate;
exports.isUndef = isUndef;
exports.isUndefNull = isUndefNull;
exports.isDef = isDef;
exports.isTrue = isTrue;
exports.isFalse = isFalse;
exports.isNull = isNull;
exports.isPromise = isPromise;
exports.isRegExp = isRegExp;

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

/**
 * 获取变量类型
 * @param v
 * @returns {string}
 * @private
 */
function _getType(v) {
  return Object.prototype.toString.call(v);
}
/**
 * 判断是否为函数
 * @param v
 * @returns {boolean}
 */


function isFunc(v) {
  return typeof v === 'function';
}
/**
 * 判断是否为异步函数
 * @param v
 * @returns {boolean}
 */


function isAsyncFunc(v) {
  return _getType(v) === '[object AsyncFunction]';
}
/**
 * 判断是否为数字
 * @param v
 * @returns {boolean}
 */


function isNum(v) {
  return _getType(v) === '[object Number]';
}
/**
 * 判断是否为数组
 * @param v
 * @returns {boolean}
 */


function isArray(v) {
  return _getType(v) === '[object Array]';
}
/**
 * 判断是否为数组
 * @param v
 * @returns {boolean}
 */


function isAry(v) {
  return isArray(v);
}
/**
 * 判断是否为布尔值
 * @param v
 * @returns {boolean}
 */


function isBoolean(v) {
  return _getType(v) === '[object Boolean]';
}
/**
 * 判断是否为布尔值
 * @param v
 * @returns {boolean}
 */


function isBol(v) {
  return isBoolean(v);
}
/**
 * 判断是否为对象
 * @param v
 * @returns {*}
 */


function isObject(v) {
  return _getType(v) === '[object Object]';
}
/**
 * 判断是否为对象
 * @param v
 * @returns {*}
 */


function isObj(v) {
  return isObject(v);
}
/**
 * 判断是否为字符串
 * @param v
 * @returns {*}
 */


function isString(v) {
  return _getType(v) === '[object String]';
}
/**
 * 判断是否为字符串
 * @param v
 * @returns {*}
 */


function isStr(v) {
  return isString(v);
}
/**
 * 判断是否为日期对象
 * @param v
 * @returns {boolean}
 */


function isDate(v) {
  return _getType(v) === '[object Date]';
}
/**
 * 判断是否未定义
 * @param v
 * @returns {boolean}
 */


function isUndef(v) {
  return v === undefined;
}
/**
 * 判断是否未定义或者null
 * @param v
 * @returns {boolean}
 */


function isUndefNull(v) {
  return isUndef(v) || isNull(v);
}
/**
 * 判断是否为定义变量
 * @param v
 * @returns {boolean}
 */


function isDef(v) {
  return v !== undefined && v !== null;
}
/**
 * 判断是否为true
 * @param v
 * @returns {boolean}
 */


function isTrue(v) {
  return v === true;
}
/**
 * 判断是否为false
 * @param v
 * @returns {boolean}
 */


function isFalse(v) {
  return v === false;
}
/**
 * 判断是否为null
 * @param v
 * @returns {boolean}
 */


function isNull(v) {
  return v === null;
}
/**
 * 判断是否为promise对象
 * @param v
 * @returns {*|"undefined"|"object"|"boolean"|"number"|"string"|"function"|"symbol"|"bigint"}
 */


function isPromise(v) {
  return isDef(v) && _typeof(isFunc(v.then)) && isFunc(v["catch"]);
}
/**
 * 判断是否为正则表达式
 * @param v
 * @returns {boolean}
 */


function isRegExp(v) {
  return _getType(v) === '[object RegExp]';
}
